File: ui/layout.go
// Copyright (c) 2026 lazybeeper by Ronen Druker.

package ui

// Layout holds calculated panel dimensions based on the terminal size.
type Layout struct {
	// TotalWidth is the terminal width.
	TotalWidth int
	// TotalHeight is the terminal height.
	TotalHeight int
	// SidebarWidth is the width of the sidebar column.
	SidebarWidth int
	// MainWidth is the width of the main content area.
	MainWidth int
	// AccountsHeight is the height of the accounts panel.
	AccountsHeight int
	// ChatsHeight is the height of the chats panel.
	ChatsHeight int
	// MessagesHeight is the height of the message viewport.
	MessagesHeight int
	// InputHeight is the height of the input textarea.
	InputHeight int
	// StatusBarHeight is the height of the status bar (always 1).
	StatusBarHeight int
}

// Sidebar width constraints.
const (
	sidebarMinWidth = 25
	sidebarMaxWidth = 40
	sidebarPct      = 30
	inputHeight     = 5
	statusBarHeight = 1
)

// CalculateLayout computes panel dimensions from the total terminal size.
func CalculateLayout(width, height int) Layout {
	lay := Layout{
		TotalWidth:      width,
		TotalHeight:     height,
		StatusBarHeight: statusBarHeight,
	}

	// Sidebar width: 30% clamped to [25, 40].
	sideW := width * sidebarPct / 100
	if sideW < sidebarMinWidth {
		sideW = sidebarMinWidth
	}

	if sideW > sidebarMaxWidth {
		sideW = sidebarMaxWidth
	}

	lay.SidebarWidth = sideW
	lay.MainWidth = width - sideW

	// Vertical layout: subtract status bar.
	usableHeight := height - statusBarHeight
	if usableHeight < 0 {
		usableHeight = 0
	}

	// Sidebar: accounts 30%, chats 70%.
	lay.AccountsHeight = usableHeight * 30 / 100
	if lay.AccountsHeight < 3 {
		lay.AccountsHeight = 3
	}

	lay.ChatsHeight = usableHeight - lay.AccountsHeight

	// Main area: input is fixed, messages get the rest.
	lay.InputHeight = inputHeight
	lay.MessagesHeight = usableHeight - lay.InputHeight

	if lay.MessagesHeight < 3 {
		lay.MessagesHeight = 3
		lay.InputHeight = usableHeight - lay.MessagesHeight

		if lay.InputHeight < 3 {
			lay.InputHeight = 3
		}
	}

	return lay
}

File: src/ui/layout.ts
// Copyright (c) 2026 lazybeeper by Ronen Druker.

/** Minimum sidebar width in characters. */
const SIDEBAR_MIN_WIDTH = 25;

/** Maximum sidebar width in characters. */
const SIDEBAR_MAX_WIDTH = 40;

/** Sidebar width as a percentage of total width. */
const SIDEBAR_PCT = 30;

/** Fixed height for the input panel. */
const INPUT_HEIGHT = 6;

/** Fixed height for the status bar. */
const STATUS_BAR_HEIGHT = 1;

/**
 * Layout holds calculated panel dimensions based on the terminal size.
 */
export interface Layout {
  /** Total terminal width. */
  readonly totalWidth: number;
  /** Total terminal height. */
  readonly totalHeight: number;
  /** Width of the sidebar column. */
  readonly sidebarWidth: number;
  /** Width of the main content area. */
  readonly mainWidth: number;
  /** Height of the accounts panel. */
  readonly accountsHeight: number;
  /** Height of the chats panel. */
  readonly chatsHeight: number;
  /** Height of the message viewport. */
  readonly messagesHeight: number;
  /** Height of the input textarea. */
  readonly inputHeight: number;
  /** Height of the status bar (always 1). */
  readonly statusBarHeight: number;
}

/**
 * Computes panel dimensions from the total terminal size.
 * @param width - The total terminal width in characters.
 * @param height - The total terminal height in characters.
 * @returns The computed layout dimensions.
 */
export function calculateLayout(width: number, height: number): Layout {
  let sidebarWidth = Math.floor((width * SIDEBAR_PCT) / 100);
  sidebarWidth = Math.max(SIDEBAR_MIN_WIDTH, Math.min(SIDEBAR_MAX_WIDTH, sidebarWidth));

  const mainWidth = width - sidebarWidth;

  let usableHeight = height - STATUS_BAR_HEIGHT;
  if (usableHeight < 0) {
    usableHeight = 0;
  }

  let accountsHeight = Math.floor((usableHeight * 30) / 100);
  if (accountsHeight < 3) {
    accountsHeight = 3;
  }

  const chatsHeight = usableHeight - accountsHeight;

  let inputHeight = INPUT_HEIGHT;
  let messagesHeight = usableHeight - inputHeight;

  if (messagesHeight < 3) {
    messagesHeight = 3;
    inputHeight = usableHeight - messagesHeight;

    if (inputHeight < 3) {
      inputHeight = 3;
    }
  }

  return {
    totalWidth: width,
    totalHeight: height,
    sidebarWidth,
    mainWidth,
    accountsHeight,
    chatsHeight,
    messagesHeight,
    inputHeight,
    statusBarHeight: STATUS_BAR_HEIGHT,
  };
}
